### PR TITLE
gnome-base/gnome-shell: Add ebuild with musl fix

### PR DESCRIPTION
**Pull request by vadorovsky**
The HAVE_EXE_INTROSPECTION variable is enabled if elf.h and link.h headers exist, but then the code relies on the _DYNAMIC variable which exists only in glibc link.h.

### CODE REVIEW

File: gnome-base/gnome-shell/files/44.1-Disable-HAVE_EXE_INTROSPECTION.patch [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+diff --git a/meson.build b/meson.build
+index 7b08fe52d..5ec06d371 100644
+--- a/meson.build
++++ b/meson.build
+@@ -148,8 +148,7 @@ cdata.set('HAVE_FDWALK', cc.has_function('fdwalk'))
+ cdata.set('HAVE_MALLINFO', cc.has_function('mallinfo'))
+ cdata.set('HAVE_MALLINFO2', cc.has_function('mallinfo2'))
+ cdata.set('HAVE_SYS_RESOURCE_H', cc.has_header('sys/resource.h'))
+-cdata.set('HAVE_EXE_INTROSPECTION',
```
Comment (thesamesam) [inline gnome-base/gnome-shell/files/44.1-Disable-HAVE_EXE_INTROSPECTION.patch:20]: Can you send this upstream with a patch for it to check for `_DYNAMIC` properly (or guard it entirely on glibc if it's using internals)?

Comment (thesamesam) [reply gnome-base/gnome-shell/files/44.1-Disable-HAVE_EXE_INTROSPECTION.patch:20]: Hm, see https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2747. They seem to think it's fine on musl even.

Comment (vadorovsky) [reply gnome-base/gnome-shell/files/44.1-Disable-HAVE_EXE_INTROSPECTION.patch:20]: Sure, I will try to fix that upstream.

Comment (thesamesam) [reply gnome-base/gnome-shell/files/44.1-Disable-HAVE_EXE_INTROSPECTION.patch:20]: I'm not sure why you closed this, I was asking for two things:
1. How this squares with https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2747 which implies this shouldn't be a problem
1. For you to send it upstream and include a link in the patch

Comment (vadorovsky) [reply gnome-base/gnome-shell/files/44.1-Disable-HAVE_EXE_INTROSPECTION.patch:20]: Sorry for misunderstanding.

https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2747 implies that executable introspection works on Linux with musl, but that isn't true (at least from my experience). On my Gentoo with musl+LLVM profile, building gnome-shell fails with:

```
[266/280] Compiling C object src/gnome-shell.p/main.c.o
FAILED: src/gnome-shell.p/main.c.o
cc -Isrc/gnome-shell.p -Isrc -I../src -I. -I.. -Isrc/st -I../src/st -Isrc/tray -I../src/tray -I/usr/include/gio-unix-2.0 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/lib/libffi/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/libxml2 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -I/usr/include/atk-1.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/gjs-1.0 -I/usr/include/gobject-introspection-1.0 -I/usr/include/nspr -I/usr/include/cairo -I/usr/include/freetype2 -I/usr/include/pixman-1 -I/usr/include/mutter-12/clutter -I/usr/include/json-glib-1.0 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/fribidi -I/usr/include/mutter-12/cogl -I/usr/include/graphene-1.0 -I/usr/lib/graphene-1.0/include -I/usr/include/startup-notification-1.0 -I/usr/include/polkit-1 -I/usr/include/gcr-4 -I/usr/include/p11-kit-1 -I/usr/include/gck-2 -I/usr/include/libnm -I/usr/include/libsecret-1 -I/usr/include/mutter-12 -I/usr/include/gsettings-desktop-schemas -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -fno-strict-aliasing -Wpointer-arith -Wmissing-declarations -Wimplicit-function-declaration -Wformat=2 -Wformat-nonliteral -Wformat-security -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wold-style-definition -Wundef -Wunused -Wcast-align -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-include-dirs -Wlogical-op -Wignored-qualifiers -Werror=redundant-decls -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -Werror=empty-body -Werror=write-strings -DG_ENABLE_DEBUG -fno-omit-frame-pointer -pthread -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -isystem /usr/include/mozjs-102 -DCLUTTER_ENABLE_EXPERIMENTAL_API -DCOGL_ENABLE_EXPERIMENTAL_API '-DVERSION="44.3"' '-DLOCALEDIR="/usr/local/share/locale"' '-DDATADIR="/usr/local/share"' '-DGNOME_SHELL_LIBEXECDIR="/usr/local/libexec"' '-DGNOME_SHELL_DATADIR="/usr/local/share/gnome-shell"' '-DGNOME_SHELL_PKGLIBDIR="/usr/local/lib/gnome-shell"' '-DMUTTER_TYPELIB_DIR="/usr/lib/mutter-12"' -MD -MQ src/gnome-shell.p/main.c.o -MF src/gnome-shell.p/main.c.o.d -o src/gnome-shell.p/main.c.o -c ../src/main.c
../src/main.c: In function 'maybe_add_rpath_introspection_paths':
../src/main.c:143:14: error: '_DYNAMIC' undeclared (first use in this function); did you mean 'PT_DYNAMIC'?
  143 |   for (dyn = _DYNAMIC; dyn->d_tag != DT_NULL; dyn++)
      |              ^~~~~~~~
      |              PT_DYNAMIC
../src/main.c:143:14: note: each undeclared identifier is reported only once for each function it appears in
[273/280] Generating src/st/St-12.gir with a custom command (wrapped by meson to set env)
ninja: build stopped: subcommand failed.
```

And I took your comment as "please prepare a patch to gnome-shell, which guards the `_DYNAMIC` part on glibc, include a musl-compliant alternative for the guarded code, send it upstream, then come back here with the same patch". Which made sense for me, so I closed this PR, which just disables the whole feature entirely (like the upstream PR 2747 does for non-Linux systems). And unless you have any other suggestions, that's what I'm planning to do.

Also, sorry for late reaction, I will do my best to be more responsive and work on the proper fix now.

**Comment by vadorovsky** [comment]
For anyone interested in running GNOME with musl - I gave up on GNOME entirely (for reasons not related to this PR) and switched to KDE, which builds with LLVM+musl without any issues, except this little patch for Dolphin (gentoo/musl#484) which is upstreamed. I guess Plasma 6 is not going to need any patches, yay!

I still think the patch from this PR is needed to build GNOME Shell (but I didn't try since 2 months). This condition is going to be resolved to "true" even with musl systems (it has both `elf.h` and `link.h` headers, but without GNU extensions):

https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/meson.build?ref_type=heads#L167-168

Which results in compiling this code, which relies on GNU extension (`_DYNAMIC`, not present in musl's link.h):

https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/src/main.c#L149

**Pull request closed without merging**